src/librustc_mir/mir_map.rs

```diff
@@ -227,15 +225,16 @@ fn closure_self_ty<'a, 'tcx>(tcx: &ty::ctxt<'tcx>,
     let region = tcx.mk_region(region);
 
     match tcx.closure_kind(tcx.map.local_def_id(closure_expr_id)) {
-        ty::ClosureKind::FnClosureKind =>
-            tcx.mk_ref(region,
-                       ty::TypeAndMut { ty: closure_ty,
-                                        mutbl: hir::MutImmutable }),
-        ty::ClosureKind::FnMutClosureKind =>
-            tcx.mk_ref(region,
-                       ty::TypeAndMut { ty: closure_ty,
-                                        mutbl: hir::MutMutable }),
-        ty::ClosureKind::FnOnceClosureKind =>
-            closure_ty
+        ty::ClosureKind::FnClosureKind => tcx.mk_ref(region,
```

Comment:
Perhaps that same heuristic should also be applied to `=>` in patterns (i.e., consider a break there). I find this pretty cramped. @nrc @marcusklaas